src/lib/tls/tls12/msg_server_kex.cpp

```diff
@@ -97,33 +97,23 @@ Server_Key_Exchange::Server_Key_Exchange(Handshake_IO& io,
          throw TLS_Exception(Alert::HandshakeFailure, "No shared ECC group with client");
       }
 
-      std::vector<uint8_t> ecdh_public_val;
-
-      if(m_shared_group.value() == Group_Params::X25519 || m_shared_group.value() == Group_Params::X448) {
-         m_kex_key = state.callbacks().tls_generate_ephemeral_key(m_shared_group.value(), rng);
-         if(!m_kex_key) {
-            throw TLS_Exception(Alert::InternalError, "Application did not provide an EC key");
-         }
-         ecdh_public_val = m_kex_key->public_value();
-      } else {
-         m_kex_key = state.callbacks().tls_generate_ephemeral_key(m_shared_group.value(), rng);
-         auto* ecdh = dynamic_cast<ECDH_PrivateKey*>(m_kex_key.get());
-         if(ecdh == nullptr) {
-            throw TLS_Exception(Alert::InternalError, "Application did not provide a EC-Diffie-Hellman key");
-         }
-
-         // follow client's preference for point compression
-         ecdh_public_val =
-            ecdh->public_value(state.client_hello()->prefers_compressed_ec_points() ? EC_Point_Format::Compressed
-                                                                                    : EC_Point_Format::Uncompressed);
+      const auto pubkey_point_format = state.client_hello()->prefers_compressed_ec_points()
+                                          ? EC_Point_Format::Compressed
+                                          : EC_Point_Format::Uncompressed;
+      m_kex_key = state.callbacks().tls_generate_ephemeral_ecdh_key(m_shared_group.value(), rng, pubkey_point_format);
+      if(!m_kex_key) {
+         throw TLS_Exception(Alert::InternalError, "Application did not provide an EC key");
       }
 
       const uint16_t named_curve_id = m_shared_group.value().wire_code();
       m_params.push_back(3);  // named curve
       m_params.push_back(get_byte<0>(named_curve_id));
       m_params.push_back(get_byte<1>(named_curve_id));
 
-      append_tls_length_value(m_params, ecdh_public_val, 1);
+      // Note: In contrast to public_value(), raw_public_key_bits() takes the
+      // point format (compressed vs. uncompressed) into account that was set
+      // in its construction within tls_generate_ephemeral_key().
+      append_tls_length_value(m_params, m_kex_key->raw_public_key_bits(), 1);
```

Comment:
Is there any other failsafe, when the `tls_generate_ephemeral_ecdh_key()`  ignores the point format? TLS may fail, right?